myia/parser.py

```diff
@@ -0,0 +1,958 @@
+
+            for block in function.blocks:
+                if not block.used:
+                    continue
+                repl = {}
+                repl_seq = {}
+
+                # resolve all variable reads
+                for var, items in block.variables_read.items():
+                    for item in items:
+                        self.resolve_read(
+                            repl, repl_seq, item, function, local_namespace
+                        )
+
+                # resolve write that need to stay
+                for var, items in block.variables_written.items():
+                    for item in items:
+                        self.resolve_write(
+                            repl, repl_seq, item, function, local_namespace
+                        )
+
+                # make sure to "bake in" top level chain replacements
+                # replacements inside the subgraphs are handled by replace
+                for k in list(repl_seq):
+                    n = repl_seq[k]
+                    while n in repl_seq or n in repl:
+                        if n in repl_seq:
+                            n = repl_seq[n]
+                        else:
+                            n = repl[n]
+                    repl_seq[k] = n
+
+                for k in list(repl):
+                    n = repl[k]
+                    while n in repl:  # pragma: nocover
+                        assert (
+                            False
+                        ), "Please report the code that triggered this"
+                        n = repl[n]
+                    repl[k] = n
+
+                block.graph.replace(repl, repl_seq)
+
+    def _process_args(self, block, function_block, args):
+        pargs = args.args
+        nondefaults = [None] * (len(pargs) - len(args.defaults))
+        defaults = nondefaults + args.defaults
+
+        kwargs = args.kwonlyargs
+        kwnondefaults = [None] * (len(kwargs) - len(args.kw_defaults))
+        kwdefaults = kwnondefaults + args.kw_defaults
+
+        defaults_name = []
+        defaults_list = []
+
+        # We don't handle this for now.
+        assert args.posonlyargs == []
+
+        for arg, dflt in zip(pargs + kwargs, defaults + kwdefaults):
+            with debug_inherit(name=arg.arg, location=self.make_location(arg)):
+                param_node = Parameter(function_block.graph, name=arg.arg)
+
+            if arg.annotation:
+                param_node.add_annotation(self._eval_ast_node(arg.annotation))
+
+            function_block.graph.parameters.append(param_node)
+            function_block.write(arg.arg, param_node)
+            if dflt:
+                # TODO If there are no parents (the initial function),
+                # then evaluate in the global env
+                if block is None:
+                    raise MyiaSyntaxError(
+                        "default value on the entry function",
+                        self.make_location(arg),
+                    )
```

Comment:
Could we simply use `_eval_ast_node` here?

Reply:
Maybe, but I don't think it would work because that might need a bunch of context that we don't have.

I would prefer to leave it like that for the merge, but we can revisit after.